airflow/contrib/hooks/opsgenie_alert_hook.py

```diff
@@ -0,0 +1,173 @@
+# under the License.
+#
+
+import json
+
+from airflow.hooks.http_hook import HttpHook
+from airflow.exceptions import AirflowException
+
+
+class OpsgenieAlertHook(HttpHook):
+    """
+    This hook allows you to post alerts to Opsgenie.
+    Accepts a connection that has an Opsgenie API key as the connection's password.
+    Each Opsgenie API key can be pre-configured to a team integration.
+    You can override these defaults in this hook.
+
+    :param http_conn_id: Http connection ID with host as "https://api.opsgenie.com/"
+                      and Opsgenie API key as the connection's password
+                      (e.g. "eb243592-faa2-4ba2-a551q-1afdf565c889")
+    :type http_conn_id: str
+    :param message: The Message of the Opsgenie alert
+    :type message: str
+    :param alias: Client-defined identifier of the alert
+    :type alias: str
+    :param description: Description field of the alert
+    :type description: str
+    :param responders: Teams, users, escalations and schedules that
+                      the alert will be routed to send notifications.
+    :type responders: list[dict]
+    :param visible_to: Teams and users that the alert will become visible
+                      to without sending any notification.
+    :type visible_to: list[dict]
+    :param actions: Custom actions that will be available for the alert.
+    :type actions: list[str]
+    :param tags: Tags of the alert.
+    :type tags: list[str]
+    :param details: Map of key-value pairs to use as custom properties of the alert.
+    :type details: dict
+    :param entity: Entity field of the alert that is
+                    generally used to specify which domain alert is related to.
+    :type entity: str
+    :param source: Source field of the alert. Default value is
+                    IP address of the incoming request.
+    :type source: str
+    :param priority: Priority level of the alert. Default value is P3.
+    :type priority: str
+    :param user: Display name of the request owner.
+    :type user: str
+    :param note: Additional note that will be added while creating the alert.
+    :type note: str
+    :param proxy: Proxy to use to make the Opsgenie Alert API call
+    :type proxy: str
+    """
+    def __init__(self,
+                 http_conn_id=None,
+                 message="",
+                 alias=None,
+                 description=None,
+                 responders=None,
+                 visible_to=None,
+                 actions=None,
+                 tags=None,
+                 details=None,
+                 entity=None,
+                 source=None,
+                 priority=None,
+                 user=None,
+                 note=None,
+                 proxy=None,
+                 *args,
+                 **kwargs
+                 ):
+        super(OpsgenieAlertHook, self).__init__(*args, **kwargs)
+        self.http_conn_id = http_conn_id
+        self.api_key = self._get_api_key(http_conn_id)
```

Comment:
This will cause a performance problem. The operator is created every few seconds while parsing DAG. Your code will cause a new database query to be called every few seconds to read the connection data, but in most cases this value will not be used anywhere. The constructor should be light and not create any new objects. Currently my team has a similar problem with operators for GCP. In these operators, we create a hook in the constructor.
CC: @zhongjiajie

Reply:
Moved out of constructor

Reply:
Agree with @mik-laj point, good catch